crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2059,6 +2114,293 @@ impl<'db> TypeInferenceBuilder<'db> {
+                self.infer_generator_expression(generator);
+                Type::Unknown
+            }
+            ast::Expr::Await(await_expression) => {
+                self.infer_await_expression(await_expression);
+                Type::Unknown
+            }
+            ast::Expr::Yield(yield_expression) => {
+                self.infer_yield_expression(yield_expression);
+                Type::Unknown
+            }
+            ast::Expr::YieldFrom(yield_from) => {
+                self.infer_yield_from_expression(yield_from);
+                Type::Unknown
+            }
+            ast::Expr::Compare(compare) => {
+                self.infer_compare_expression(compare);
+                Type::Unknown
+            }
+            ast::Expr::Call(call_expr) => {
+                self.infer_call_expression(call_expr);
+                Type::Unknown
+            }
+            ast::Expr::FString(fstring) => {
+                self.infer_fstring_expression(fstring);
+                Type::Unknown
+            }
+            ast::Expr::Attribute(attribute) => {
+                self.infer_attribute_expression(attribute);
+                Type::Unknown
+            }
+            // TODO: this may be a place we need to revisit with special forms.
+            ast::Expr::Subscript(subscript) => {
+                self.infer_subscript_expression(subscript);
+                Type::Unknown
+            }
+            ast::Expr::Starred(starred) => {
+                self.infer_starred_expression(starred);
+                Type::Unknown
+            }
+            ast::Expr::List(list) => {
+                self.infer_list_expression(list);
+                Type::Unknown
+            }
+            ast::Expr::Tuple(tuple) => {
+                self.infer_tuple_expression(tuple);
+                Type::Unknown
+            }
+            ast::Expr::Slice(slice) => {
+                self.infer_slice_expression(slice);
+                Type::Unknown
+            }
+
+            // Literals do not have meaningful values in the annotation expression context. However,
+            // we will we want to handle these differently when working with special forms, since
+            // `"hello"` is not valid in an annotation expression but `Literal["hello"]` is.
+            ast::Expr::StringLiteral(_literal) => Type::Unknown,
+            ast::Expr::BytesLiteral(_literal) => Type::Unknown,
+            ast::Expr::NumberLiteral(_literal) => Type::Unknown,
+            ast::Expr::BooleanLiteral(_literal) => Type::Unknown,
+            ast::Expr::EllipsisLiteral(_literal) => Type::Unknown,
+
+            ast::Expr::IpyEscapeCommand(_) => todo!("Implement Ipy escape command support"),
+        };
+
+        let expr_id = expression.scoped_ast_id(self.db, self.scope);
+        tracing::trace!(
+            expr_id = expr_id.as_u32(),
+            ty = ty.display(self.db).to_string(),
+        );
+        self.types.expressions.insert(expr_id, ty);
```

Comment:
Given that we will in the future have standalone type expressions, I think it's important that `infer_type_expression` insert the expression type, so I think this should also move down to `infer_type_expression`.

For now I don't think we need it in `infer_annotation_expression` at all (in keeping with the above comment about making it a very light wrapper around `infer_type_expression`). In future it should insert _something_ into the map for whatever annotation-only special forms it handles; we can figure that out when the time comes to handle those special forms.

Reply:
Ah, that makes sense, yes. Will do. 👍🏼